src/edu/washington/escience/myria/operator/Sample.java

```diff
@@ -0,0 +1,241 @@
+      if (ans.hasFilledTB()) {
+        return ans.popFilled();
+      }
+    }
+    return ans.popAny();
+  }
+
+  /** Helper function to extract sampling information from a TupleBatch. */
+  private void extractSamplingInfo(TupleBatch tb) throws Exception {
+    Preconditions.checkArgument(tb != null);
+
+    int workerID;
+    Type col0Type = tb.getSchema().getColumnType(0);
+    if (col0Type == Type.INT_TYPE) {
+      workerID = tb.getInt(0, 0);
+    } else if (col0Type == Type.LONG_TYPE) {
+      workerID = (int) tb.getLong(0, 0);
+    } else {
+      throw new DbException("WorkerID column must be of type INT or LONG");
+    }
+    Preconditions.checkState(workerID == getNodeID(),
+            "Invalid WorkerID for this worker. Expected %s, but received %s",
+            getNodeID(), workerID);
+
+    Type col1Type = tb.getSchema().getColumnType(1);
+    if (col1Type == Type.INT_TYPE) {
+      streamSize = tb.getInt(1, 0);
+    } else if (col1Type == Type.LONG_TYPE) {
+      streamSize = (int) tb.getLong(1, 0);
+    } else {
+      throw new DbException("StreamSize column must be of type INT or LONG");
+    }
+    Preconditions.checkState(streamSize >= 0, "streamSize cannot be negative");
+
+    Type col2Type = tb.getSchema().getColumnType(2);
+    if (col2Type == Type.INT_TYPE) {
+      sampleSize = tb.getInt(2, 0);
+    } else if (col2Type == Type.LONG_TYPE) {
+      sampleSize = (int) tb.getLong(2, 0);
+    } else {
+      throw new DbException("SampleSize column must be of type INT or LONG");
+    }
+    Preconditions.checkState(sampleSize >= 0, "sampleSize cannot be negative");
+
+    Type col3Type = tb.getSchema().getColumnType(3);
+    if (col3Type == Type.STRING_TYPE) {
+      sampleType = tb.getString(3, 0);
+    } else {
+      throw new DbException("SampleType column must be of type STRING");
+    }
+  }
+
+  /**
+   * Generates a sorted array of random numbers to be taken as samples.
+   *
+   * @param populationSize
+   *          size of the population that will be sampled from.
+   * @param sampleSize
+   *          number of samples to draw from the population.
+   * @return a sorted array of indices.
+   */
+  private int[] generateIndicesWR(int populationSize, int sampleSize) {
+    int[] indices = new int[sampleSize];
+    for (int i = 0; i < sampleSize; i++) {
+      indices[i] = rand.nextInt(populationSize);
+    }
+    Arrays.sort(indices);
+    return indices;
+  }
+
+  /**
+   * Generates a sorted array of unique random numbers to be taken as samples.
+   *
+   * @param populationSize
+   *          size of the population that will be sampled from.
+   * @param sampleSize
+   *          number of samples to draw from the population.
+   * @return a sorted array of indices.
+   */
+  private int[] generateIndicesWoR(int populationSize, int sampleSize) {
+    Set<Integer> indices = new HashSet<Integer>(sampleSize);
+    for (int i = populationSize - sampleSize; i < populationSize; i++) {
```

Comment:
This looks like Floyd's algorithm, in which case you should give a reference, since it's sufficiently non-obvious to likely cause puzzlement.